analysis/analysis.go

```diff
@@ -0,0 +1,254 @@
+				metricResult.Measurements[len(metricResult.Measurements)-1] = newMeasurement
+			}
+			if newMeasurement.Status.Completed() {
+				switch newMeasurement.Status {
+				case v1alpha1.AnalysisStatusSuccessful:
+					metricResult.Successful++
+					metricResult.Count++
+				case v1alpha1.AnalysisStatusFailed:
+					metricResult.Failed++
+					metricResult.Count++
+				case v1alpha1.AnalysisStatusInconclusive:
+					metricResult.Inconclusive++
+					metricResult.Count++
+				case v1alpha1.AnalysisStatusError:
+					metricResult.Error++
+				}
+			}
+		}(nil, task)
+		//}(provider, task)
+	}
+	wg.Wait()
+}
+
+// asssessRunStatus assesses the overall status of this AnalysisRun
+// If any metric is not yet completed, the AnalysisRun is still considered Running
+// Once all metrics are complete, the worst status is used as the overall AnalysisRun status
+func asssessRunStatus(run *v1alpha1.AnalysisRun) v1alpha1.AnalysisStatus {
+	var worstStatus v1alpha1.AnalysisStatus
+	terminating := analysisutil.IsTerminating(run)
+	for _, metric := range run.Spec.AnalysisSpec.Metrics {
+		if result, ok := run.Status.MetricResults[metric.Name]; ok {
+			metricStatus := assessMetricStatus(&metric, &result, terminating)
+			if !metricStatus.Completed() {
+				// if any metric is not completed, then entire analysis run is considered running
+				return v1alpha1.AnalysisStatusRunning
+			}
+			if worstStatus == "" {
+				worstStatus = metricStatus
+			} else {
+				if analysisutil.IsWorse(worstStatus, metricStatus) {
+					worstStatus = metricStatus
+				}
+			}
+		}
+	}
+	return worstStatus
+}
+
+// assessMetricStatus assesses the status of a single metric based on:
+// * current/latest measurement status
+// * parameters given by the metric (maxFailures, count, etc...)
+// * whether or not we are terminating (e.g. due to failing run, or termination request)
+func assessMetricStatus(metric *v1alpha1.Metric, result *v1alpha1.MetricResult, terminating bool) v1alpha1.AnalysisStatus {
+	if result.Status.Completed() {
+		return result.Status
+	}
+	log := log.WithField("metric", metric.Name)
+	if len(result.Measurements) == 0 {
+		if terminating {
+			// we have yet to take a single measurement, but have already been instructed to stop
+			log.Infof("metric assessed %s: run terminated", v1alpha1.AnalysisStatusSuccessful)
+			return v1alpha1.AnalysisStatusSuccessful
+		}
+		return v1alpha1.AnalysisStatusPending
+	}
+	lastMeasurement := result.Measurements[len(result.Measurements)-1]
+	if !lastMeasurement.Status.Completed() {
+		// we still have a in-flight measurement
+		return v1alpha1.AnalysisStatusRunning
+	}
+	if result.Failed > metric.MaxFailures {
+		log.Infof("metric assessed %s: failed (%d) > maxFailures (%d)", v1alpha1.AnalysisStatusFailed, result.Failed, metric.MaxFailures)
+		return v1alpha1.AnalysisStatusFailed
+	}
+	consecutiveErrors := analysisutil.ConsecutiveErrors(result)
+	if consecutiveErrors > MaxConsecutiveErrors {
```

Comment:
I was implementing the default behavior. I think there does need to be some default behavior because user's wont think to specify this.

I will add a maxConsecutiveErrors field to change from default.